presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -586,4 +596,165 @@ public void prepareFinal()
+            }
+
+            Block[] blocks = new Block[spillingTypes.size()];
+            for (int channel = 0; channel < spillingTypes.size(); channel++) {
+                blocks[channel] = columnarRow.getField(channel);
+            }
+            Page page = new Page(blocks);
+            GroupByIdBlock squashedGroupIds = new GroupByIdBlock(groupIdsBlock.getGroupCount(), new LongArrayBlock(newPositionCount, Optional.of(nulls), newGroupIds));
+
+            rawInputsSize += page.getSizeInBytes();
+            rawInputs.add(new GroupIdPage(squashedGroupIds, page));
+        }
+
+        @Override
+        public void evaluateIntermediate(int groupId, BlockBuilder output)
+        {
+            if (blockBuilders == null) {
+                if (rawInputs == null) {
+                    return;
+                }
+                blockBuilders = new ObjectBigArray<>();  // we need to spill now; let's build the hash table here
+                for (GroupIdPage groupIdPage : rawInputs) {
+                    Page page = groupIdPage.getPage();
+                    GroupByIdBlock groupIdsBlock = groupIdPage.getGroupByIdBlock();
+                    for (int position = 0; position < page.getPositionCount(); position++) {
+                        long currentGroupId = groupIdsBlock.getGroupId(position);
+                        blockBuilders.ensureCapacity(currentGroupId);
+                        RowBlockBuilder rowBlockBuilder = blockBuilders.get(currentGroupId);
+                        if (rowBlockBuilder == null) {
+                            rowBlockBuilder = new RowBlockBuilder(spillingTypes, null, (int) groupIdsBlock.getGroupCount());
+                        }
+
+                        BlockBuilder currentOutput = rowBlockBuilder.beginBlockEntry();
+                        for (int channel = 0; channel < spillingTypes.size(); channel++) {
+                            spillingTypes.get(channel).appendTo(page.getBlock(channel), position, currentOutput);
+                        }
+                        rowBlockBuilder.closeEntry();
+
+                        blockBuilders.set(currentGroupId, rowBlockBuilder);
+                    }
+                }
+                rawInputs = null;
+                rawInputsSize = 0;
+            }
+
+            BlockBuilder singleArrayBlockWriter = output.beginBlockEntry();
+            checkState(rawInputs == null && blockBuilders != null);
+            checkState(singleArrayBlockWriter instanceof SingleArrayBlockWriter);
+
+            RowBlock rowBlock = (RowBlock) blockBuilders.get((long) groupId).build();
+            for (int i = 0; i < rowBlock.getPositionCount(); i++) {
+                singleArrayBlockWriter.appendStructure(rowBlock.getBlock(i));
+            }
+            output.closeEntry();
+        }
+
+        @Override
+        public void evaluateFinal(int groupId, BlockBuilder output)
+        {
+            // checkState(rawInputs == null && blockBuilders == null);
```

Comment:
Do we know if there will be a correctness issue if an aggregationBuilder already starts to build result and we start to revoke and produce intermediate ressult?

Reply:
Take a look at `hasBuiltFinalResult` in `InMemoryHashAggregationBuilder` - that may fix it but unsure about the implications.